File: library/src/main/java/com/yy/codex/uikit/UINavigationBar.java
package com.yy.codex.uikit;

import android.app.Activity;
import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.os.Build;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.RequiresApi;
import android.util.AttributeSet;
import android.view.View;

import com.yy.codex.foundation.NSInvocation;

import java.util.HashMap;

/**
 * Created by cuiminghui on 2017/1/18.
 */

public class UINavigationBar extends UIView {

    public UINavigationBar(@NonNull Context context, @NonNull View view) {
        super(context, view);
    }

    public UINavigationBar(@NonNull Context context) {
        super(context);
    }

    public UINavigationBar(@NonNull Context context, @NonNull AttributeSet attrs) {
        super(context, attrs);
    }

    public UINavigationBar(@NonNull Context context, @NonNull AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
    public UINavigationBar(@NonNull Context context, @NonNull AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
    }

    @Override
    protected void init() {
        super.init();
        UIConstraint constraint = new UIConstraint();
        constraint.centerHorizontally = true;
        constraint.top = "0";
        constraint.width = "100%";
        constraint.height = "44";
        setConstraint(constraint);
    }

    @Override
    public void willMoveToSuperview(@Nullable UIView newSuperview) {
        super.willMoveToSuperview(newSuperview);
        setBackgroundColor(getBarTintColor());
        setTintColor(getTintColor());
    }

    @Override
    public void didMoveToSuperview() {
        super.didMoveToSuperview();
        if (getContext() != null && getContext() instanceof Activity) {
            NSInvocation invocation = new NSInvocation(getContext(), "getSupportActionBar");
            Object actionBar = invocation.invoke();
            if (actionBar != null) {
                NSInvocation actionBarInvocation = new NSInvocation(actionBar, "hide");
                actionBarInvocation.invoke();
            }
            else {
                NSInvocation _invocation = new NSInvocation(getContext(), "getActionBar");
                Object _actionBar = _invocation.invoke();
                if (_actionBar != null) {
                    NSInvocation _actionBarInvocation = new NSInvocation(_actionBar, "hide");
                    _actionBarInvocation.invoke();
                }
            }
        }
    }

    /* Material Design */

    private boolean mMaterialDesignInitialized = false;

    @Override
    public void materialDesignDidChanged() {
        super.materialDesignDidChanged();
        if (!mMaterialDesignInitialized && isMaterialDesign()) {
            mMaterialDesignInitialized = true;
            setBarTintColor(new UIColor(0x3f/255.0, 0x51/255.0, 0xb5/255.0, 1.0));
            setTintColor(UIColor.whiteColor);
            setTitleTextAttributes(new HashMap<String, Object>(){{
                put(NSAttributedString.NSForegroundColorAttributeName, UIColor.whiteColor);
                put(NSAttributedString.NSFontAttributeName, UIFont.systemBold(17));
            }});
            if (getConstraint() != null) {
                getConstraint().height = "48";
            }
            layoutSubviews();
            resetItemsView();
            invalidate();
        }
    }

    /* BarTintColor */

    private UIColor mBarTintColor;

    public UIColor getBarTintColor() {
        if (mBarTintColor == null) {
            mBarTintColor = new UIColor(0xf8/255.0, 0xf8/255.0, 0xf8/255.0, 1.0);
        }
        return mBarTintColor;
    }

    public void setBarTintColor(UIColor barTintColor) {
        mBarTintColor = barTintColor;
        setBackgroundColor(barTintColor);
    }

    /* Title Attributes */

    @Nullable private HashMap<String, Object> mTitleTextAttributes;

    @Nullable
    public HashMap<String, Object> getTitleTextAttributes() {
        return mTitleTextAttributes;
    }

    public void setTitleTextAttributes(@Nullable HashMap<String, Object> titleTextAttributes) {
        mTitleTextAttributes = titleTextAttributes;
        resetItemsView();
    }

    /* Line Color */

    private UIColor mBottomLineColor;

    public UIColor getBottomLineColor() {
        if (mBottomLineColor == null) {
            mBottomLineColor = new UIColor(0xb2/255.0, 0xb2/255.0, 0xb2/255.0, 0.5);
        }
        return mBottomLineColor;
    }

    public void setBottomLineColor(UIColor bottomLineColor) {
        mBottomLineColor = bottomLineColor;
        invalidate();
    }

    @Override
    protected void onDraw(@NonNull Canvas canvas) {
        super.onDraw(canvas);
        if (!isMaterialDesign()) {
            Paint paint = new Paint();
            paint.setColor(getBottomLineColor().toInt());
            canvas.drawLine(0, canvas.getHeight() - 1, canvas.getWidth(), canvas.getHeight() - 1, paint);
        }
    }

    /* Items */

    private UINavigationItem[] mItems = new UINavigationItem[0];
    private UINavigationItemView[] mItemsView = new UINavigationItemView[0];

    public void setItems(UINavigationItem[] items, boolean animated) {
        mItems = items;
        resetItemsProps();
        resetBackItems();
        resetItemsView();
    }

    public void pushNavigationItem(UINavigationItem item, boolean animated) {
        UINavigationItem[] items = new UINavigationItem[mItems.length + 1];
        for (int i = 0; i < mItems.length; i++) {
            items[i] = mItems[i];
        }
        items[items.length - 1] = item;
        mItems = items;
        resetItemsProps();
        resetBackItems();
        resetItemsView();
        if (animated) {
            doPushAnimation();
        }
    }

    protected void doPushAnimation() {
        if (mItemsView.length >= 2) {
            final UINavigationItemView topItemView = mItemsView[mItemsView.length - 1];
            final UINavigationItemView backItemView = mItemsView[mItemsView.length - 2];
            topItemView.setAlpha(1);
            backItemView.setAlpha(1);
            topItemView.animateToFront(false);
            backItemView.animateFromFrontToBack(false);
            UIView.animator.linear(0.75, new Runnable() {
                @Override
                public void run() {
                    topItemView.animateToFront(true);
                    backItemView.animateFromFrontToBack(true);
                }
            }, new Runnable() {
                @Override
                public void run() {
                    backItemView.setAlpha(0);
                }
            });
        }
    }

    public void popNavigationItem(boolean animated) {
        if (animated) {
            doPopAnimation(new Runnable() {
                @Override
                public void run() {
                    popNavigationItem(false);
                }
            });
        }
        else {
            if (mItems.length <= 0) {
                return;
            }
            UINavigationItem[] items = new UINavigationItem[mItems.length - 1];
            for (int i = 0; i < mItems.length; i++) {
                if (i < items.length) {
                    items[i] = mItems[i];
                }
            }
            mItems = items;
            resetItemsView();
        }
    }

    protected void doPopAnimation(@NonNull final Runnable completion) {
        if (mItemsView.length >= 2) {
            final UINavigationItemView topItemView = mItemsView[mItemsView.length - 1];
            final UINavigationItemView backItemView = mItemsView[mItemsView.length - 2];
            topItemView.setAlpha(1);
            backItemView.setAlpha(1);
            topItemView.animateToGone(false);
            backItemView.animateFromBackToFront(false);
            UIView.animator.linear(0.75, new Runnable() {
                @Override
                public void run() {
                    topItemView.animateToGone(true);
                    backItemView.animateFromBackToFront(true);
                }
            }, new Runnable() {
                @Override
                public void run() {
                    topItemView.setAlpha(0);
                    if (completion != null) {
                        completion.run();
                    }
                }
            });
        }
    }

    protected void resetItemsProps() {
        for (int i = 0; i < mItems.length; i++) {
            mItems[i].setNavigationBar(this);
        }
    }

    protected void resetBackItems() {
        for (int i = 0; i < mItems.length; i++) {
            if (i > 0) {
                if (mItems[i].getLeftBarButtonItems().length <= 0 || mItems[i].getLeftBarButtonItems()[0].isSystemBackItem()) {
                    mItems[i].setLeftBarButtonItem(mItems[i - 1].getBackBarButtonItem());
                }
            }
        }
    }

    protected void resetItemsView() {
        for (int i = 0; i < mItemsView.length; i++) {
            mItemsView[i].removeFromSuperview();
        }
        UINavigationItemView[] itemsView = new UINavigationItemView[mItems.length];
        for (int i = 0; i < mItems.length; i++) {
            UINavigationItemView frontView = isMaterialDesign() ? new UINavigationItemView_MaterialDesign(getContext()) : new UINavigationItemView(getContext());
            mItems[i].mFrontView = frontView;
            mItems[i].setNeedsUpdate();
            itemsView[i] = frontView;
            frontView.setConstraint(UIConstraint.full());
            addSubview(frontView);
            if (i < mItems.length - 1) {
                frontView.setAlpha(0);
            }
            else {
                frontView.setAlpha(1);
            }
        }
        mItemsView = itemsView;
        layoutSubviews();
    }

}

File: library/src/main/java/com/yy/codex/uikit/UINavigationBar.kt
package com.yy.codex.uikit

import android.app.Activity
import android.content.Context
import android.graphics.Canvas
import android.graphics.Paint
import android.os.Build
import android.support.annotation.RequiresApi
import android.util.AttributeSet
import android.view.View

import com.yy.codex.foundation.NSInvocation

import java.util.HashMap

/**
 * Created by cuiminghui on 2017/1/18.
 */

class UINavigationBar : UIView {

    constructor(context: Context, view: View) : super(context, view) {}

    constructor(context: Context) : super(context) {}

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {}

    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {}

    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int, defStyleRes: Int) : super(context, attrs, defStyleAttr, defStyleRes) {}

    override fun init() {
        super.init()
        val constraint = UIConstraint()
        constraint.centerHorizontally = true
        constraint.top = "0"
        constraint.width = "100%"
        constraint.height = "44"
        this.constraint = constraint
    }

    override fun willMoveToSuperview(newSuperview: UIView?) {
        super.willMoveToSuperview(newSuperview)
        setBackgroundColor(barTintColor)
    }

    override fun didMoveToSuperview() {
        super.didMoveToSuperview()
        if (context != null && context is Activity) {
            val invocation = NSInvocation(context, "getSupportActionBar")
            val actionBar = invocation.invoke()
            if (actionBar != null) {
                val actionBarInvocation = NSInvocation(actionBar, "hide")
                actionBarInvocation.invoke()
            } else {
                val _invocation = NSInvocation(context, "getActionBar")
                val _actionBar = _invocation.invoke()
                if (_actionBar != null) {
                    val _actionBarInvocation = NSInvocation(_actionBar, "hide")
                    _actionBarInvocation.invoke()
                }
            }
        }
    }

    /* Layout Length */

    fun length(): Double {
        if (isMaterialDesign) {
            return 48.0
        }
        return 44.0
    }

    /* Material Design */

    private var mMaterialDesignInitialized = false

    override fun materialDesignDidChanged() {
        super.materialDesignDidChanged()
        if (!mMaterialDesignInitialized && isMaterialDesign) {
            mMaterialDesignInitialized = true
            barTintColor = UIColor(0x3f / 255.0, 0x51 / 255.0, 0xb5 / 255.0, 1.0)
            tintColor = UIColor.whiteColor
            titleTextAttributes = object : HashMap<String, Any>() {
                init {
                    put(NSAttributedString.NSForegroundColorAttributeName, UIColor.whiteColor)
                    put(NSAttributedString.NSFontAttributeName, UIFont.systemBold(17f))
                }
            }
            if (constraint != null) {
                constraint!!.height = "48"
            }
            layoutSubviews()
            resetItemsView()
            invalidate()
        }
    }

    /* BarTintColor */

    var barTintColor: UIColor = UIColor(0xf8 / 255.0, 0xf8 / 255.0, 0xf8 / 255.0, 1.0)
        set(barTintColor) {
            field = barTintColor
            setBackgroundColor(barTintColor)
        }

    /* Title Attributes */

    var titleTextAttributes: HashMap<String, Any>? = null
        set(titleTextAttributes) {
            field = titleTextAttributes
            resetItemsView()
        }

    /* Line Color */

    var bottomLineColor: UIColor = UIColor(0xb2 / 255.0, 0xb2 / 255.0, 0xb2 / 255.0, 0.5)
        set(bottomLineColor) {
            field = bottomLineColor
            invalidate()
        }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        if (!isMaterialDesign) {
            val paint = Paint()
            paint.color = bottomLineColor.toInt()
            canvas.drawLine(0f, (canvas.height - 1).toFloat(), canvas.width.toFloat(), (canvas.height - 1).toFloat(), paint)
        }
    }

    /* Items */

    private var items: List<UINavigationItem> = listOf()
    private var itemsView: List<UINavigationItemView> = listOf()

    fun setItems(items: List<UINavigationItem>, animated: Boolean) {
        this.items = items
        resetItemsProps()
        resetBackItems()
        resetItemsView()
    }

    fun pushNavigationItem(item: UINavigationItem, animated: Boolean) {
        val mutableItems = items.toMutableList()
        mutableItems.add(item)
        this.items = mutableItems.toList()
        resetItemsProps()
        resetBackItems()
        resetItemsView()
        if (animated) {
            doPushAnimation()
        }
    }

    fun doPushAnimation() {
        if (itemsView.size >= 2) {
            val topItemView = itemsView[itemsView.size - 1]
            val backItemView = itemsView[itemsView.size - 2]
            topItemView.alpha = 1f
            topItemView.animateToFront(false)
            backItemView.alpha = 1f
            backItemView.animateFromFrontToBack(false)
            UIViewAnimator.linear(0.75, Runnable {
                topItemView.animateToFront(true)
                backItemView.animateFromFrontToBack(true)
            }, Runnable { backItemView.alpha = 0f })
        }
    }

    fun popNavigationItem(animated: Boolean) {
        if (animated) {
            doPopAnimation(Runnable { popNavigationItem(false) })
        } else {
            val mutableItems = items.toMutableList()
            mutableItems.removeAt(mutableItems.count() - 1)
            this.items = mutableItems.toList()
            resetItemsView()
        }
    }

    fun doPopAnimation(completion: Runnable) {
        if (itemsView.size >= 2) {
            val topItemView = itemsView[itemsView.size - 1]
            val backItemView = itemsView[itemsView.size - 2]
            topItemView.alpha = 1f
            backItemView.alpha = 1f
            topItemView.animateToGone(false)
            backItemView.animateFromBackToFront(false)
            UIViewAnimator.linear(0.75, Runnable {
                topItemView.animateToGone(true)
                backItemView.animateFromBackToFront(true)
            }, Runnable {
                topItemView.alpha = 0f
                completion.run()
            })
        }
    }

    fun resetItemsProps() {
        for (item in items) {
            item.navigationBar = this
        }
    }

    fun resetBackItems() {
        for ((index, item) in items.withIndex()) {
            if (index > 0) {
                if (item.leftBarButtonItems.count() <= 0 || item.leftBarButtonItems.first().isSystemBackItem) {
                    item.setLeftBarButtonItem(items[index - 1].backBarButtonItem)
                }
            }
            else {
                if (item.leftBarButtonItems.count() == 1 && item.leftBarButtonItems.first().isSystemBackItem) {
                    item.setLeftBarButtonItem(null)
                }
            }
        }
    }

    fun resetItemsView() {
        for (itemView in itemsView) {
            itemView.removeFromSuperview()
        }
        val theItemsView: MutableList<UINavigationItemView> = mutableListOf()
        for ((index, item) in items.withIndex()) {
            val frontView = if (isMaterialDesign) UINavigationItemView_MaterialDesign(context) else UINavigationItemView(context)
            item.frontView = frontView
            item.setNeedsUpdate()
            theItemsView.add(frontView)
            frontView.constraint = UIConstraint.full()
            addSubview(frontView)
            if (index < items.count() - 1) {
                frontView.alpha = 0f
            }
            else {
                frontView.alpha = 1f
            }
        }
        this.itemsView = theItemsView
        layoutSubviews()
    }

}
